### ISSUE TITLE
why bbox is merged?

### ISSUE THREAD

**Issue by YoungjaeDev** [ultralytics/yolov5#9985]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello
You will be able to check the issue I raised by referring to the attached photo below
The content is that each individual box is merged into one box
I usually detect Boat, and I looked into the model I trained for a problem, but I saw the same issue in the pretrained-coco model

I checked the issue with the command below. And the confidence threshold has been increased significantly, so can I know the cause and solution?

```bash
yolov5 detect --weights yolov5m.pt --source ../../track01.mp4 --imgsz 960 --max-det 300 --device 0 --conf-thres 0.45
```

Thanks 

### Additional

![image](https://user-images.githubusercontent.com/99321359/198954046-695cde4f-5e12-4984-9070-d47d4af3ea7a.png)
![image](https://user-images.githubusercontent.com/99321359/198954158-2ab451b2-6ac8-4ef4-a210-bd338727d121.png)
![image](https://user-images.githubusercontent.com/99321359/198954296-054013e8-ec7d-4aee-8870-dc53c8d21852.png)
![image](https://user-images.githubusercontent.com/99321359/198955154-88aa9dda-d571-45d5-958b-b4c9ab74ab4c.png)

**Comment by glenn-jocher** [comment]
@youngjae-avikus you can update detection settings in detect.py to suit your requirements, i.e. --conf --iou --agnostic, etc.

**Comment by youngjae-avikus** [comment]
> @youngjae-avikus you can update detection settings in detect.py to suit your requirements, i.e. --conf --iou --agnostic, etc.

Yes, that's right.  However, the merge seems to have nothing to do with the parameters below.  detecting individual boats and multiple boats should coexist, but that's not the case.  Could it be that there is a way to suppress this during model training or because the test image quality is poor?

**Comment by MartinPedersenpp** [comment]
> > @youngjae-avikus you can update detection settings in detect.py to suit your requirements, i.e. --conf --iou --agnostic, etc.
> 
> @glenn-jocher Yes, that's right. However, the merge seems to have nothing to do with the parameters below. detecting individual boats and multiple boats should coexist, but that's not the case. Could it be that there is a way to suppress this during model training or because the test image quality is poor?

IoU indeed has something to do with it, if two boxes with the same label are overlapping, the IoU will decide whether they get merged or left as two boxes.

**Comment by youngjae-avikus** [comment]
@MartinPedersenpp 
Well, I didn't give you a clear idea in the first place. I think it's strange to hold several boats in one box in the first place. In addition, isn't iou irrelevant to merge?

**Comment by MartinPedersenpp** [comment]
https://github.com/ultralytics/yolov5/blob/a83d2a50132982fa89a22420155f6c9f097a92c7/utils/general.py#L845-L962

When running inference, the IoU is used as a weight for merging bboxes, compared to when doing evaluation it is used for calculating how close the prediction is to the ground truth.
I don't have the degree to explain it in more detail, but there are plenty of articles out there if you google how IoU works
From my experience, a low Iou will force boxes to be merged

**Comment by youngjae-avikus** [comment]
> https://github.com/ultralytics/yolov5/blob/a83d2a50132982fa89a22420155f6c9f097a92c7/utils/general.py#L845-L962
> When running inference, the IoU is used as a weight for merging bboxes, compared to when doing evaluation it is used for calculating how close the prediction is to the ground truth. I don't have the degree to explain it in more detail, but there are plenty of articles out there if you google how IoU works From my experience, a low Iou will force boxes to be merged

Isn't merge false by default?

**Comment by MartinPedersenpp** [comment]
@youngjae-avikus 

Yes but as default, Yolo NMS uses Torchvison.ops.nms (https://pytorch.org/vision/stable/generated/torchvision.ops.nms.html)

**Comment by youngjae-avikus** [comment]
@MartinPedersenpp 

I think it's no different from original nms in terms of explanation, do you mean that merge can be operated internally based on weight in `Torchvison.ops.nms` ?

**Comment by MartinPedersenpp** [comment]
> @MartinPedersenpp
> 
> I think it's no different from original nms in terms of explanation, do you mean that merge can be operated internally based on weight in `Torchvison.ops.nms` ?

`iou_threshold – discards all overlapping boxes with IoU > iou_threshold`

So basically with the default IoU at 0.45, your boats bboxes are overlapping more than the threshold, hence, the smallest is removed.

Try passing --iou 0.95 to your detect, this way you can almost see all of the bboxes predicted by your model.

**Comment by youngjae-avikus** [comment]
thanks for your reply
 However, the fact that the box that holds the whole is maintained is still a problem.

**Comment by MartinPedersenpp** [comment]
> thanks for your reply However, the fact that the box that holds the whole is maintained is still a problem.

If the detection itself is wrong, then your options are to either give a input with a better resolution/contrast between the boats, so the model can tell them apart, unless I open and zoom in on the problematic image, I have a hard time telling them apart as well(I am on a 14" monitor)
Alternatively, you can train your model with some more images of boats overlapping and mess around with hyperparameters during training to make up for the varying size of the boats.
A last option could be to set IoU to 0.95 and write a custom algorithm for filtering your bboxes eg.
````
if bbox__inliers > 1:
    remove biggest_bbox
    keep n bbox_inliers
````

I hope this helps, good luck!

**Comment by youngjae-avikus** [comment]
@MartinPedersenpp 

I think it's because I don't have a lot of various data. The algorithm you suggested is good, but there are also edge cases with ships covered in large ships, so it will not be easy to use. 
Thank you.
